Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -1,36 +1,70 @@
 #include "BatteryStateOfChargeService.h"
+#include <QDebug>
 
 namespace
 {
     const double BATTERY_AMP_HOUR_CAPACITY = 123.0;
 }
 
 BatteryStateOfChargeService::BatteryStateOfChargeService(double initialStateOfChargePercent)
-: initialStateOfChargePercent_(initialStateOfChargePercent)
+    : initialStateOfChargePercent_(initialStateOfChargePercent)
 {
 }
 
+
 BatteryStateOfChargeService::~BatteryStateOfChargeService()
 {
 }
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return initialAhUsed_ + chargeAh_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (current_ < 0)
+       return true;
+        return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime time(0,0,0,0);
+    int ms_remaining;
+      if (isCharging() == true)  {
+        ms_remaining = totalAmpHoursUsed()/averageCurrent_ *3.6E6;
+    }   else    {
+        ms_remaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed())/averageCurrent_ *3.6E6;
+        ms_remaining *= -1;
+    }
+
+    QTime timeWhenChargedOrDepleted = time.addMSecs(ms_remaining);
+
+    return timeWhenChargedOrDepleted;
+
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
+
+    QTime updated_time;
+
+    double DeltaMseconds = batteryData.time.msec() - Previous_milliseconds_.msec();
```

Comment:
One more point to consider, `DeltaMseconds` relies on the value of `Previous_milliseconds.msec()`, however when on the very first run, `Previous.millseconds.msec()` doesn't have a value at all, so the results can be skewed. I'll let you try and figure out how to work around this.